gettsim/tax_transfer.py

```diff
@@ -56,47 +46,200 @@ def tax_transfer(df, tb, tb_pens=None):
+    # df[column] = 0.0
+    household = ["hid"]
+    tax_unit = ["hid", "tu_id"]
+    person = ["hid", "tu_id", "pid"]
     # We start with the top layer, which is household id. We treat this as the
-    # "Bedarfsgemeinschaft" in the german tax law.
+    # "Bedarfsgemeinschaft" in German tax law.
+    in_cols = [
+        "m_wage",
+        "east",
+        "age",
+        "selfemployed",
+        "haskids",
+        "m_self",
+        "m_pensions",
+        "pkv",
+    ]
+    out_cols = ["svbeit", "rvbeit", "avbeit", "gkvbeit", "pvbeit"]
+    for col in out_cols:
+        df[col] = np.nan
+    df = df.groupby(person)[in_cols + out_cols].apply(soc_ins_contrib, tb=tb)
+    in_cols = [
+        "m_wage_l1",
+        "east",
+        "child",
+        "months_ue",
+        "months_ue_l1",
+        "months_ue_l2",
+        "alg_soep",
+        "m_pensions",
+        "w_hours",
+        "child_num_tu",
+        "age",
+    ]
+    out_col = "m_alg1"
+    df[out_col] = np.nan
+    df = df.groupby(person)[in_cols + [out_col]].apply(ui, tb=tb)
+    in_cols = ["m_wage", "east", "age", "year", "byear", "exper", "EP"]
+    out_col = "pensions_sim"
+    df[out_col] = np.nan
+    df = df.groupby(person)[in_cols + [out_col]].apply(pensions, tb=tb)
+    in_cols = [
+        "m_wage",
+        "m_self",
+        "m_kapinc",
+        "m_vermiet",
+        "renteneintritt",
+        "m_pensions",
+        "w_hours",
+        "ineducation",
+        "zveranl",
+        "child",
+        "m_childcare",
+        "handcap_degree",
+        "rvbeit",
+        "avbeit",
+        "pvbeit",
+        "alleinerz",
+        "age",
+        "child_num_tu",
+        "year",
+        "east",
+        "gkvbeit",
+    ]
+    out_cols = [
+        "zve_nokfb",
+        "zve_abg_nokfb",
+        "zve_kfb",
+        "zve_abg_kfb",
+        "kifreib",
+        "gross_e1",
+        "gross_e4",
+        "gross_e5",
+        "gross_e6",
+        "gross_e7",
+        "gross_e1_tu",
+        "gross_e4_tu",
+        "gross_e5_tu",
+        "gross_e6_tu",
+        "gross_e7_tu",
+        "ertragsanteil",
+        "sonder",
+        "hhfreib",
+        "altfreib",
+        "vorsorge",
+    ]
+    df = df.groupby(tax_unit)[in_cols + out_cols].apply(zve, tb=tb)
+    in_cols = ["age", "w_hours", "ineducation", "m_wage"]
+    out_cols = ["kindergeld_basis", "kindergeld_tu_basis"]
+    for col in out_cols:
+        df[col] = np.nan
+    df = df.groupby(tax_unit)[in_cols + out_cols].apply(kindergeld, tb=tb)
+    in_cols = [
+        "zveranl",
+        "child",
+        "tax_nokfb_tu",
+        "tax_kfb_tu",
+        "abgst_tu",
+        "kindergeld_basis",
+        "kindergeld_tu_basis",
+    ]
+    out_cols = [
+        "incometax_tu",
+        "incometax",
+        "kindergeld",
+        "kindergeld_hh",
+        "kindergeld_tu",
+    ]
+    for col in out_cols:
+        df[col] = np.nan
+    df = df.groupby(tax_unit)[in_cols + out_cols].apply(favorability_check, tb=tb)
```

Comment:
This whole sequence is highly repetitive. Can be possibly put in a loop / function as well.

Reply:
I wrote a function, which does that for me.